servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/WriteStreamSubscriber.java

```diff
@@ -420,6 +482,38 @@ private void terminateSubscriber(@Nullable Throwable cause) {
             }
         }
 
+        private void notifyAllListeners(@Nullable Throwable cause) {
+            final ChannelFuture future = cause == null ? channel.newSucceededFuture() : channel.newFailedFuture(cause);
+            while (!listenersOnWriteBoundaries.isEmpty()) {
+                Object mayBeListener = listenersOnWriteBoundaries.removeFirst();
+                if (mayBeListener != WRITE_BOUNDARY) {
+                    notifyListener(future, mayBeListener);
+                }
+            }
+        }
+
+        private void notifyListenersTillNextWrite(@Nullable Throwable cause) {
+            Object shdBeWriteBoundary = listenersOnWriteBoundaries.removeFirst();
```

Comment:
Let me try this out, thanks

Reply:
Ok doing this will have problems if the write completes synchronously with `channel.write()` call.

```
listenersOnWriteBoundaries <- empty
channel.write()
   - addListener(listener1) => listenersOnWriteBoundaries (listener1)
   - writeAndFlush() (internally in the pipeline) => drain  listenersOnWriteBoundaries <- empty
add WRITE_BOUNDARY to listenersOnWriteBoundaries
channel.write()
    - addListener(listener2) => listenersOnWriteBoundaries (WRITE_BOUNDARY, listener2)
flush()
notifyListenersTillNextWrite() <- removes WRITE_BOUNDARY and exits
```

Reply:
I have replaced the early-return with an `assert`

Reply:
good call on the re-entry! Here is a unit test for `WriteStreamSubscriberFutureListenersTest` to verify correctness here:

```java
    @Test
    public void synchronousCompleteWrite() throws Exception {
        Channel mockChannel = mock(Channel.class);
        EventLoop mockEventLoop = mock(EventLoop.class);
        when(mockEventLoop.inEventLoop()).thenReturn(true);
        when(mockChannel.eventLoop()).thenReturn(mockEventLoop);
        when(mockChannel.newSucceededFuture()).thenReturn(channel.newSucceededFuture());
        doAnswer((Answer<Void>) invocation -> {
            ReferenceCountUtil.release(invocation.getArgument(0));
            ChannelFutureListener listener = mock(ChannelFutureListener.class);
            listeners.add(listener);
            ChannelPromise promise = invocation.getArgument(1);
            promise.addListener(listener);
            promise.setSuccess();
            return null;
        }).when(mockChannel).write(any(), any());
        WriteDemandEstimator estimator = WriteDemandEstimators.newDefaultEstimator();
        TestCompletableSubscriber completableSubscriber = new TestCompletableSubscriber();
        WriteStreamSubscriber subscriber = new WriteStreamSubscriber(mockChannel, estimator, completableSubscriber,
                UNSUPPORTED_PROTOCOL_CLOSE_HANDLER);
        subscriber.onNext(1);
        verifyListenerInvokedWithSuccess(listeners.take());
        subscriber.onNext(2);
        verifyListenerInvokedWithSuccess(listeners.take());
    }
```